Fix slow backtracking when parsing strings (no external deps)

TysonAndre opened the pull request:
Fixes #61

This uses negative lookaheads to avoid ambiguity in how string should be
parsed by the regex.
- https://docs.python.org/2/library/re.html#regular-expression-syntax

- Previously, if it didn't immediately succeed at parsing an escape
  sequence such as `\123`, it would have to try `\1`+`23`, `\12` + `3`,
  and `\123`, which multiplied the time taken by 3 per additional escape
  sequence. This solves that by only allowing `\123`
- The same fix was added for hex escapes.

Also fix a test that relied on the incorrect handling of regexes.
The implementation documentation says that it intends to allow
**decimal** escapes permissively.

TysonAndre commented:
This still has backtracking for `"\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\`\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x23\x2\x23456` without a closing quote. Investigating.

TysonAndre commented:
In the previous commits,
`\x12` could be parsed both as `\x`+`12` and `\x12`,
which caused exponential options for backtracking, because "\a-z" was allowed for windows paths.

I solved this by allowing "\x" when not followed by hex
in the regular string literal (and treating \x followed by at least one hex character as hex)

eliben commented on pycparser/c_lexer.py:
```diff
@@ -205,10 +205,10 @@ def _make_tok_location(self, token):
-    decimal_escape = r"""(\d+)"""
-    hex_escape = r"""(x[0-9a-fA-F]+)"""
-    bad_escape = r"""([\\][^a-zA-Z._~^!=&\^\-\\?'"x0-7])"""
+    simple_escape = r"""([a-wyzA-Z._~!=&\^\-\\?'"]|x(?![0-9a-fA-F]))"""
```
Please add a sizable comment explaining how these regexes are structured and why. The original regexes were taken verbatim from the C syntax definition, IIRC

eliben commented on pycparser/ply/lex.py:
```diff
@@ -317,7 +317,13 @@ def token(self):
             # Look for a regular expression match
             for lexre, lexindexfunc in self.lexre:
-                m = lexre.match(lexdata, lexpos)
+                try:
```
Debugging leftovers?

TysonAndre replied:
Yes, they are. Removed.

TysonAndre commented:
> Thanks for pursuing this. I'm thinking about the added complexity vs. performance win in edge cases. It's not clear to me it's worth it, yet, to be honest. One of the things I'm considering is whether this reduces performance in more "regular" use cases, but unfortunately pycparser doesn't have a benchmark suite.

Added another simplified regex just for strings (but not character literals, which expect just one escape sequence). The unit tests I added continue to pass.

This takes advantage of the fact that `\[0-9]` is valid on its own, and `\x` is valid on its own, due to the permissive nature, and we don't need the following sequences of non-escaped characters themselves.

eliben commented:
Could you please run the benchmark script I added in https://github.com/eliben/pycparser/tree/master/utils/benchmark without and then with your changes applied and report the results here?

TysonAndre commented:
1. In general, benchmarking results on my laptop have been all over the place for various projects. I'd assumed this was due to mitigations for intel prefetching bugs, power saving, etc., but I'm not actually sure
2. The script fails when run on utils/benchmark/inputs due to a ParseError on .gitattributes. I removed that to run the script

Example results are included below, for origin/master and my branch

```
±3b89659⚡ » taskset 1 python3 utils/benchmark/benchmark-parse.py utils/benchmark/inputs 
tccgen.c.ppout           .....    Mean: 0.860  Stddev: 0.014
sqlite-btree.c.ppout     .....    Mean: 1.385  Stddev: 0.012
redis.c.ppout            .....    Mean: 0.586  Stddev: 0.040
±3b89659⚡ » taskset 1 python3 utils/benchmark/benchmark-parse.py utils/benchmark/inputs 
tccgen.c.ppout           .....    Mean: 1.061  Stddev: 0.031
sqlite-btree.c.ppout     .....    Mean: 1.502  Stddev: 0.028
redis.c.ppout            .....    Mean: 0.527  Stddev: 0.016
±3b89659⚡ » taskset 1 python3 utils/benchmark/benchmark-parse.py utils/benchmark/inputs 
tccgen.c.ppout           .....    Mean: 0.873  Stddev: 0.012
sqlite-btree.c.ppout     .....    Mean: 1.407  Stddev: 0.011
redis.c.ppout            .....    Mean: 0.604  Stddev: 0.032
```

```
±regex-fix-backtracking-v2⚡ » taskset 1 python3 utils/benchmark/benchmark-parse.py utils/benchmark/inputs 
tccgen.c.ppout           .....    Mean: 0.952  Stddev: 0.133
sqlite-btree.c.ppout     .....    Mean: 1.383  Stddev: 0.004
redis.c.ppout            .....    Mean: 0.543  Stddev: 0.052
±regex-fix-backtracking-v2⚡ » taskset 1 python3 utils/benchmark/benchmark-parse.py utils/benchmark/inputs 
tccgen.c.ppout           .....    Mean: 0.868  Stddev: 0.014
sqlite-btree.c.ppout     .....    Mean: 1.389  Stddev: 0.005
redis.c.ppout            .....    Mean: 0.531  Stddev: 0.055
```
